sosw/components/sns.py

```diff
@@ -141,22 +152,74 @@ def commit(self):
         message = self.separator.join(self.queue)
 
         if message:
+            extra_args = {}
+
+            # Add MessageAttributes to the message if they are defined
+            if self.message_attributes:
+                extra_args['MessageAttributes'] = {
+                    key: self.get_message_attribute(value) for key, value in self.message_attributes.items()
+                }
+
+            logger.info(f"MessageAttributes: {self.message_attributes}")
+
             self.client.publish(
                     TopicArn=self.recipient,
                     Subject=self.subject,
-                    Message=message)
+                    Message=message,
+                    **extra_args
+            )
 
         self.queue = []
+        self.message_attributes = None
+
 
+    def compare_message_attributtes(self, message_attributes):
+        """
+        Compare MessageAttributes. If new object was received - set new value to the attribute
 
-    def send_message(self, message, subject=None, forse_commit=False):
+        :param dict message_attributes: Message attributes for SNS subscription filter policies
+        """
+
+        if not self.message_attributes == message_attributes:
+            logger.info("Attribute message_attributes was changed. We set a new one and commit the current queue.")
+            self.set_message_attributes(message_attributes)
+
+
+    @staticmethod
+    def get_message_attribute(attribute):
+        """
+        Amazon SNS compares policy attributes only to message attributes that have the following data types:
+        - String
+        - String.Array
+        - Number
+
+        :return: Type and value of an attribute
+        :rtype: dict
+        """
+
+        if isinstance(attribute, str):
+            return {'DataType': 'String', 'StringValue': attribute}
+
+        if isinstance(attribute, (int, float)):
+            return {'DataType': 'Number', 'StringValue': str(attribute)}
+
+        if hasattr(attribute, '__iter__'):
+            return {'DataType': 'String.Array', 'StringValue': json.dumps(attribute)}
+
+        raise TypeError(f"Unsupported message_attribute value was passed. Must be one of str, int, float, or iterable. "
+                        f"Got {type(attribute)}")
+
+
+    def send_message(self, message, subject=None, message_attributes=None, forse_commit=False):
```

Comment:
Changing the signature order is not good. Please double check that nobody used (*args) when calling in at least 4 repositories (including sosw itself).